src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +117,388 @@ public class ConsensusManager
+            }
+
+            this.logger.LogTrace("(-):'{0}'", newHeaders?.Consumed);
+            return newHeaders?.Consumed;
+        }
+
+        /// <summary>
+        /// A peer was disconnected, clean all its claimed headers that are above our consensus tip.
+        /// Processes any remaining blocks to download.
+        /// </summary>
+        /// <param name="peerId">The peer that is providing the headers.</param>
+        public void PeerDisconnected(int peerId)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(peerId), peerId);
+
+            lock (this.treeLock)
+            {
+                this.chainedHeaderTree.PeerDisconnected(peerId);
+                this.blockPuller.PeerDisconnected(peerId);
+                this.ProcessDownloadQueueLocked();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A callback that is triggered when a block that <see cref="ConsensusManager"/> requested was downloaded.
+        /// </summary>
+        private void ProcessDownloadedBlock(ChainedHeaderBlock chainedHeaderBlock)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            bool partialValidationRequired = false;
+
+            lock (this.treeLock)
+            {
+                partialValidationRequired = this.chainedHeaderTree.BlockDataDownloaded(chainedHeaderBlock.ChainedHeader, chainedHeaderBlock.Block);
+            }
+
+            if (partialValidationRequired)
+            {
+                this.blockValidator.StartPartialValidation(chainedHeaderBlock, this.OnPartialValidationCompletedCallback);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private void OnPartialValidationCompletedCallback(ChainedHeaderBlock chainedHeaderBlock, bool success)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:{3})", nameof(chainedHeaderBlock), chainedHeaderBlock, nameof(success), success);
+
+            if (success)
+            {
+                this.OnPartialValidationSucceeded(chainedHeaderBlock);
+            }
+            else
+            {
+                List<int> peersToBan;
+
+                lock (this.treeLock)
+                {
+                    peersToBan = this.chainedHeaderTree.PartialOrFullValidationFailed(chainedHeaderBlock.ChainedHeader);
+                }
+
+                foreach (int peerId in peersToBan)
+                {
+                    // TODO: ban and disconnect those peers
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private void OnPartialValidationSucceeded(ChainedHeaderBlock chainedHeaderBlock)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            bool reorged = false;
+            List<ChainedHeader> nextHeadersToValidate = null;
+
+            lock (this.reorgLock)
```

Comment:
This must be async lock.

Reply:
yeah it will come when I implement the reorg method.